File: latest/week3/workshop8_m2/W08_M02_demo.R
# ============================================================
# Workshop 8 — Module 2: plotly in R
# Integrated R Script — UNYT Tirana
# ============================================================
library(tidyverse); library(plotly); library(htmlwidgets)
dir.create("output", showWarnings = FALSE)

# ── 1. LOAD DATA ────────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)),
    year_added = year(date_added),
    primary_country = str_trim(str_extract(country, "^[^,]+")),
    duration_num = parse_number(duration))

yearly <- nf |>
  filter(!is.na(year_added), year_added >= 2015, year_added <= 2021) |>
  count(year_added, type)
movies <- yearly |> filter(type == "Movie")
tv <- yearly |> filter(type == "TV Show")

# ══════════════════════════════════════════════════════════
# APPROACH 1: ggplotly — convert existing ggplot
# ══════════════════════════════════════════════════════════

# ── 2. BASIC ggplotly ────────────────────────────────────
p_gg <- ggplot(yearly, aes(x = year_added, y = n, color = type,
    text = paste0("<b>", type, "</b><br>",
      "Year: ", year_added, "<br>",
      "Titles: ", format(n, big.mark = ",")))) +
  geom_line(linewidth = 1) + geom_point(size = 2) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal() +
  labs(title = "ggplotly: Netflix Additions by Type",
       x = "Year", y = "Titles Added", color = "Type")

p1 <- ggplotly(p_gg, tooltip = "text") |>
  layout(hovermode = "x unified",
    hoverlabel = list(bgcolor = "white", font = list(size = 11)))

saveWidget(p1, "output/01_ggplotly_basic.html", selfcontained = TRUE)

# ── 3. ggplotly WITH ANNOTATIONS ─────────────────────────
peak <- movies |> slice_max(n, n = 1)
decline <- round((1 - movies$n[movies$year_added == 2021] / peak$n) * 100)

p_annotated <- ggplot(yearly, aes(x = year_added, y = n, group = type,
    text = paste0(type, ": ", n, " titles (", year_added, ")"))) +
  geom_line(aes(color = type == "Movie"), linewidth = 1) +
  geom_point(data = movies, color = "#1565C0", size = 3) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#DDDDDD"),
    guide = "none") +
  theme_minimal() +
  labs(title = paste0("Movie additions declined ", decline, "% from peak"))

p2 <- ggplotly(p_annotated, tooltip = "text") |>
  layout(
    annotations = list(
      list(x = 2020, y = movies$n[movies$year_added == 2021],
        text = paste0("<b>–", decline, "%</b>"),
        showarrow = TRUE, arrowhead = 2, arrowcolor = "#E53935",
        font = list(color = "#E53935", size = 14),
        ax = -50, ay = -40)),
    hovermode = "x unified")

saveWidget(p2, "output/02_ggplotly_annotated.html", selfcontained = TRUE)

# ══════════════════════════════════════════════════════════
# APPROACH 2: plot_ly — native syntax, full control
# ══════════════════════════════════════════════════════════

# ── 4. NATIVE plot_ly LINE CHART ─────────────────────────
p3 <- plot_ly() |>
  add_trace(data = movies,
    x = ~year_added, y = ~n,
    type = "scatter", mode = "lines+markers",
    name = "Movie",
    line = list(color = "#1565C0", width = 3),
    marker = list(size = 8, color = "#1565C0"),
    hovertemplate = paste0(
      "<b>Movie</b><br>",
      "Year: %{x}<br>",
      "Titles: %{y:,}<extra></extra>")) |>
  add_trace(data = tv,
    x = ~year_added, y = ~n,
    type = "scatter", mode = "lines+markers",
    name = "TV Show",
    line = list(color = "#E53935", width = 1.5, dash = "dot"),
    marker = list(size = 5, color = "#E53935"),
    hovertemplate = paste0(
      "<b>TV Show</b><br>",
      "Year: %{x}<br>",
      "Titles: %{y:,}<extra></extra>")) |>
  layout(
    title = list(text = "plot_ly: Full control over traces and tooltips"),
    xaxis = list(title = "Year"),
    yaxis = list(title = "Titles Added"),
    hovermode = "x unified",
    template = "plotly_white")

saveWidget(p3, "output/03_plotly_native.html", selfcontained = TRUE)

# ── 5. SCATTER WITH HOVER DETAILS ────────────────────────
sample <- nf |>
  filter(!is.na(year_added), !is.na(duration_num)) |>
  sample_n(min(1500, n()))

p4 <- plot_ly(sample,
  x = ~year_added, y = ~duration_num,
  color = ~type,
  colors = c(Movie = "#1565C0", "TV Show" = "#E53935"),
  type = "scatter", mode = "markers",
  marker = list(size = 5, opacity = 0.4),
  text = ~paste0("<b>", title, "</b><br>",
    "Type: ", type, "<br>",
    "Year: ", year_added, "<br>",
    "Duration: ", duration, "<br>",
    "Rating: ", rating, "<br>",
    "Country: ", primary_country),
  hoverinfo = "text") |>
  layout(
    title = "Scatter: hover for full title metadata",
    xaxis = list(title = "Year Added"),
    yaxis = list(title = "Duration (min/seasons)"),
    hovermode = "closest")

saveWidget(p4, "output/04_scatter_details.html", selfcontained = TRUE)

# ── 6. DROPDOWN FILTER (updatemenus) ────────────────────
# Ratings filter
ratings <- nf |> filter(!is.na(rating)) |>
  count(rating, sort = TRUE) |> slice_max(n, n = 6) |> pull(rating)

rating_yearly <- nf |>
  filter(!is.na(year_added), year_added >= 2015, year_added <= 2021,
    rating %in% ratings) |>
  count(year_added, rating)

p5 <- plot_ly()
for (r in ratings) {
  sub <- rating_yearly |> filter(rating == r)
  p5 <- p5 |> add_trace(data = sub,
    x = ~year_added, y = ~n,
    type = "scatter", mode = "lines+markers",
    name = r, visible = TRUE,
    hovertemplate = paste0("<b>", r, "</b><br>Year: %{x}<br>Count: %{y}<extra></extra>"))
}

# Dropdown buttons
buttons <- list(
  list(method = "restyle",
    args = list("visible", as.list(rep(TRUE, length(ratings)))),
    label = "All Ratings"))
for (i in seq_along(ratings)) {
  vis <- rep(FALSE, length(ratings)); vis[i] <- TRUE
  buttons[[i + 1]] <- list(method = "restyle",
    args = list("visible", as.list(vis)),
    label = ratings[i])
}

p5 <- p5 |> layout(
  title = "Dropdown Filter: Select a Rating",
  updatemenus = list(list(type = "dropdown", x = 0.1, y = 1.15,
    buttons = buttons)),
  xaxis = list(title = "Year"), yaxis = list(title = "Titles"))

saveWidget(p5, "output/05_dropdown_filter.html", selfcontained = TRUE)

# ── 7. SUBPLOT: TWO CHARTS, SHARED X-AXIS ───────────────
p_line <- plot_ly(yearly, x = ~year_added, y = ~n, color = ~type,
  colors = c(Movie = "#1565C0", "TV Show" = "#E53935"),
  type = "scatter", mode = "lines+markers") |>
  layout(yaxis = list(title = "Count"))

p_bar <- plot_ly(yearly, x = ~year_added, y = ~n, color = ~type,
  colors = c(Movie = "#1565C0", "TV Show" = "#E53935"),
  type = "bar") |>
  layout(yaxis = list(title = "Count"), barmode = "group")

p6 <- subplot(p_line, p_bar, nrows = 2, shareX = TRUE, titleY = TRUE) |>
  layout(title = "subplot: line + bar with shared x-axis (zoom syncs)")

saveWidget(p6, "output/06_subplot.html", selfcontained = TRUE)

# ── 8. LINKED BRUSHING (crosstalk) ──────────────────────
library(crosstalk)
sd <- SharedData$new(sample |> filter(!is.na(rating)), key = ~show_id)

p_scatter <- plot_ly(sd, x = ~year_added, y = ~duration_num,
  color = ~type, type = "scatter", mode = "markers",
  marker = list(size = 4, opacity = 0.5))

p_hist <- plot_ly(sd, x = ~rating, type = "histogram",
  marker = list(color = "#1565C0"))

p7 <- subplot(p_scatter, p_hist, nrows = 1, widths = c(0.6, 0.4)) |>
  highlight(on = "plotly_selected", off = "plotly_deselect",
    color = "#E53935") |>
  layout(title = "Linked Brushing: select in scatter → highlights in histogram",
    dragmode = "select")

saveWidget(p7, "output/07_linked_brushing.html", selfcontained = TRUE)

cat("\n── All W08-M02 R outputs saved ──\n")
cat("7 interactive HTML files in output/\n")
cat("01: ggplotly basic\n02: ggplotly annotated\n")
cat("03: native plot_ly\n04: scatter with details\n")
cat("05: dropdown filter\n06: subplot shared axes\n")
cat("07: linked brushing (crosstalk)\n")

File: latest/week3/workshop8_m2/W08_M02_demo.py
"""W08-M02: plotly in R — Python companion showing equivalent patterns"""
import numpy as np, pandas as pd, os
os.makedirs("output", exist_ok=True)

# This module focuses on R plotly. The Python companion shows
# the EQUIVALENT plotly.py patterns for each R technique,
# enabling students to translate between the two.

try:
    import plotly.express as px
    import plotly.graph_objects as go
    from plotly.subplots import make_subplots
    HAS_PLOTLY = True
except ImportError:
    HAS_PLOTLY = False

nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
nf["primary_country"] = nf["country"].str.split(",").str[0].str.strip()
nf["duration_num"] = pd.to_numeric(nf["duration"].str.extract(r"(\d+)")[0], errors="coerce")
yearly = nf.query("2015<=year_added<=2021").groupby(["year_added","type"]).size().reset_index(name="n")
movies = yearly.query("type=='Movie'"); tv = yearly.query("type=='TV Show'")

if not HAS_PLOTLY:
    print("plotly not installed. Install with: pip install plotly")
    print("Printing R↔Python equivalence table instead.\n")

# ── 1. R↔PYTHON EQUIVALENCE TABLE ───────────────────────
print("=== R plotly ↔ Python plotly Equivalence ===\n")
equivalences = [
    ("ggplotly(p, tooltip='text')", "px.line(df, hover_data=[...])", "Quick conversion"),
    ("plot_ly(df, x=~x, y=~y)", "go.Scatter(x=df.x, y=df.y)", "Native API"),
    ("add_trace(...)", "fig.add_trace(go.Scatter(...))", "Add series"),
    ("layout(hovermode='x unified')", "fig.update_layout(hovermode=...)", "Layout config"),
    ("hovertemplate='Year: %{x}'", "hovertemplate='Year: %{x}'", "IDENTICAL syntax"),
    ("subplot(p1, p2, shareX=T)", "make_subplots(rows=2, shared_xaxes=True)", "Multi-panel"),
    ("saveWidget(p, 'f.html')", "fig.write_html('f.html')", "Save to HTML"),
    ("highlight(on='plotly_selected')", "No direct equivalent (use Dash)", "Linked brushing"),
    ("crosstalk::SharedData$new()", "No equivalent (use Dash callbacks)", "Cross-filtering"),
]
for r_code, py_code, note in equivalences:
    print(f"  R:  {r_code}")
    print(f"  Py: {py_code}")
    print(f"  ({note})\n")

if HAS_PLOTLY:
    # ── 2. px.line (equivalent to ggplotly) ──────────────
    fig1 = px.line(yearly, x="year_added", y="n", color="type",
        color_discrete_map={"Movie": "#1565C0", "TV Show": "#E53935"},
        markers=True,
        title="px.line: Python equivalent of ggplotly()",
        labels={"year_added": "Year", "n": "Titles Added", "type": "Type"})
    fig1.update_layout(hovermode="x unified", template="plotly_white")
    fig1.write_html("output/01_px_line.html")

    # ── 3. go.Scatter (equivalent to plot_ly) ────────────
    fig2 = go.Figure()
    for t, c, w in [("Movie", "#1565C0", 3), ("TV Show", "#E53935", 1.5)]:
        sub = yearly.query("type==@t")
        fig2.add_trace(go.Scatter(
            x=sub.year_added, y=sub.n,
            mode="lines+markers", name=t,
            line=dict(color=c, width=w),
            hovertemplate=f"<b>{t}</b><br>Year: %{{x}}<br>"
                f"Titles: %{{y:,}}<extra></extra>"))
    fig2.update_layout(title="go.Scatter: Python equivalent of plot_ly()",
        xaxis_title="Year", yaxis_title="Titles Added",
        hovermode="x unified", template="plotly_white")
    fig2.write_html("output/02_go_scatter.html")

    # ── 4. Scatter with hover details ────────────────────
    sample = nf.dropna(subset=["year_added","duration_num"]).sample(min(1500,len(nf)), random_state=42)
    fig3 = px.scatter(sample, x="year_added", y="duration_num",
        color="type",
        color_discrete_map={"Movie": "#1565C0", "TV Show": "#E53935"},
        hover_name="title",
        hover_data={"year_added": True, "type": True, "rating": True,
            "duration": True, "primary_country": True, "duration_num": False},
        title="Scatter with hover: full metadata on demand",
        opacity=0.4)
    fig3.update_layout(template="plotly_white", hovermode="closest")
    fig3.write_html("output/03_scatter_details.html")

    # ── 5. Dropdown filter (updatemenus) ─────────────────
    ratings = nf.dropna(subset=["rating"])["rating"].value_counts().head(6).index.tolist()
    ryr = (nf.query("2015<=year_added<=2021 and rating in @ratings")
        .groupby(["year_added","rating"]).size().reset_index(name="n"))
    fig4 = go.Figure()
    for r in ratings:
        sub = ryr.query("rating==@r")
        fig4.add_trace(go.Scatter(x=sub.year_added, y=sub.n,
            mode="lines+markers", name=r,
            hovertemplate=f"<b>{r}</b><br>Year: %{{x}}<br>Count: %{{y}}<extra></extra>"))
    buttons = [dict(label="All", method="restyle",
        args=[{"visible": [True]*len(ratings)}])]
    for i, r in enumerate(ratings):
        vis = [False]*len(ratings); vis[i] = True
        buttons.append(dict(label=r, method="restyle", args=[{"visible": vis}]))
    fig4.update_layout(title="Dropdown Filter: Select Rating",
        updatemenus=[dict(type="dropdown", x=0.1, y=1.15, buttons=buttons)],
        template="plotly_white")
    fig4.write_html("output/04_dropdown.html")

    # ── 6. Subplot with shared x-axis ────────────────────
    fig5 = make_subplots(rows=2, cols=1, shared_xaxes=True,
        subplot_titles=("Line Chart", "Bar Chart"))
    for t, c in [("Movie","#1565C0"),("TV Show","#E53935")]:
        sub = yearly.query("type==@t")
        fig5.add_trace(go.Scatter(x=sub.year_added, y=sub.n,
            mode="lines+markers", name=t, line=dict(color=c)),
            row=1, col=1)
        fig5.add_trace(go.Bar(x=sub.year_added, y=sub.n,
            name=t, marker_color=c, showlegend=False),
            row=2, col=1)
    fig5.update_layout(title="Subplot: shared x-axis (zoom syncs)",
        template="plotly_white", barmode="group")
    fig5.write_html("output/05_subplot.html")

    print("\n5 interactive HTML files saved in output/")

else:
    print("Install plotly to generate interactive HTML files.")
    print("pip install plotly")

print("\nAll W08-M02 Python outputs saved")
